catalog/core/catalog-core-camelcomponent/src/main/java/ddf/camel/component/catalog/inputtransformer/InputTransformerProducer.java

```diff
@@ -60,66 +62,71 @@ public InputTransformerProducer(CatalogEndpoint endpoint) {
     super(endpoint);
   }
 
-  @SuppressWarnings("squid:S2095" /* The TFBOS is in a try-with-resources and will be closed */)
   protected Object transform(
       Message in, String mimeType, String transformerId, MimeTypeToTransformerMapper mapper)
       throws MimeTypeParseException, CatalogTransformerException {
-    InputStream message = in.getBody(InputStream.class);
-    if (message == null) {
-      throw new CatalogTransformerException("Message body was null; unable to generate Metacard!");
-    }
-
     MimeType derivedMimeType = null;
-    if (StringUtils.isEmpty(mimeType)) {
-      try (TemporaryFileBackedOutputStream fileBackedOutputStream =
-          new TemporaryFileBackedOutputStream()) {
-        try {
-          IOUtils.copy(message, fileBackedOutputStream);
-          derivedMimeType =
-              new MimeType(
-                  getMimeTypeFromHeader(in, fileBackedOutputStream.asByteSource().openStream()));
-        } catch (IOException e) {
-          LOGGER.debug("Failed to copy incoming inputStream message", e);
-        } finally {
-          message = fileBackedOutputStream.asByteSource().openStream();
-        }
-      } catch (IOException e) {
-        LOGGER.debug("Failed to create TemporaryFileBackedOuputStream", e);
+    try (InputStream message = in.getBody(InputStream.class);
+        TemporaryFileBackedOutputStream tfbos = new TemporaryFileBackedOutputStream()) {
+      if (message == null) {
+        throw new CatalogTransformerException(
+            "Message body was null; unable to generate Metacard!");
       }
-    } else if (StringUtils.isNotBlank(transformerId)) {
-      derivedMimeType =
-          new MimeType(mimeType + ";" + MimeTypeToTransformerMapper.ID_KEY + "=" + transformerId);
-    }
 
-    if (derivedMimeType == null) {
-      derivedMimeType = new MimeType(mimeType);
-    }
+      // First try to get mimeType from file extension passed in the Camel Message headers
+      IOUtils.copy(message, tfbos);
+      String fileExtensionHeader = getHeaderAsStringAndRemove(in, FILE_EXTENSION_HEADER);
+      if (StringUtils.isNotEmpty(fileExtensionHeader)) {
+        Optional<String> fileMimeType =
+            getMimeTypeFor(tfbos.asByteSource().openBufferedStream(), fileExtensionHeader);
+        if (fileMimeType.isPresent()) {
```

Comment:
✏️ The change to Optional requires now to do a check here for presence and a get at line 87 and 85 which internally does that same check again.